Add horizontal option for the block movers

draganescu opened the pull request:
## Description
Closes #16637 

- Added the option to have horizontal mover for the blocks in the inner blocks list. 
- Changed the Navigation Menu block to use the new options.

## Known issues

- [x] Update the block description for screen readers for the horizontal situation
- [x] Should the api include position (before, after) or not?
- [x] Make the options separate instead of option objects
- [x] Handle RTL Properly:

## How has this been tested?

Tested locally by enabling the new options in the Navigation Menu block.

## Screenshots

![horizontal-movers-refactored](https://user-images.githubusercontent.com/107534/66030238-9fcd2980-e509-11e9-8a0c-301476e80b80.gif)


## Types of changes

- The new mover options are an object with two settings for now: orientation and position.
- Position determines if the movers are rendered inside (adjacent to the block edit component) or outside (in the parent of the block edit component - default).
- Orientation determines if the movers should look horizontal or vertical (default).

### Affected components
<img width="583" alt="Screenshot 2019-10-02 at 11 38 50" src="https://user-images.githubusercontent.com/107534/66030086-449b3700-e509-11e9-86eb-3b0bf6b83f99.png">


### Basic code changes
<img width="583" alt="Screenshot 2019-10-02 at 11 39 04" src="https://user-images.githubusercontent.com/107534/66030090-46fd9100-e509-11e9-969c-0b51a4816e93.png">


Props @jorgefilipecosta @talldan @jasmussen 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

draganescu committed:
horizontal option for the mover, missing icons, broken hover

draganescu committed:
we now have icons

draganescu committed:
positioned the mover to the middle left

draganescu committed:
horizontal mover on mobile

karmatosed commented:
I do wonder if moving this mover horizontal is a good step. It's taking an existing interaction model and changing to fit. That feels less than a good approach. What are we trying to solve by using it here and potentially would it be better to find a different solution?

draganescu commented:
I am unsure how to move ahead here given the state of the navigation block's UX revamp. This might be moved visually to the toolbar IDK :) Should I close this maybe?

noisysocks commented:
> I am unsure how to move ahead here given the state of the navigation block's UX revamp. This might be moved visually to the toolbar IDK :) Should I close this maybe?

Yes I'd say this is blocked pending a high level design direction for the Navigation block (https://github.com/WordPress/gutenberg/issues/16821). We can always re-open 🙂

draganescu committed:
vertical layout for horizontal movers

noisysocks commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -23,6 +23,14 @@ import useBlockNavigator from './use-block-navigator';
+>>>>>>> horizontal option for the mover, missing icons, broken hover
+=======
+	isSelected,
+>>>>>>> positioned the mover to the middle left
```
You have merge conflict markers here 🙈

draganescu replied:
ouch, I botched up something apparently ... hasty merge

draganescu committed:
drop block movers into block edit to enable inline movers

draganescu commented:
This PR now enables each Edit component to receive a BlockMover component if it is part of an InnerBlocks which has the setting horizontalMover set to true.

This allows using the horizontal movers inside the DOM of the Edit component and to make it inline etc. See PR description above for how it currently works on the Navigation block.

@mtias not sure about how I named the InnerBlocks options `horizontalMover ` ...

talldan reviewed (commented):
Left a few review comments.

jorgefilipecosta commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -402,7 +403,8 @@ function BlockListBlock( {
 		! isPartOfMultiSelection &&
-		! isTypingWithinBlock;
+		! isTypingWithinBlock &&
+		! horizontalMover;
```
It seems that shouldRenderMovers is going to be false even if it should show the horizontal movers, shouldRenderMovers should not depend if the mover is horizontal or vertical otherwise the logic is complex and it will be easy to introduce bugs in the future.

draganescu replied:
removed that with the new rendering position

jorgefilipecosta commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -511,22 +528,13 @@ function BlockListBlock( {
-						onDragEnd={ onDragEnd }
-					/>
-				) }
+				{ shouldRenderMovers && blockMover }
```
Why we only render vertical movers here? What is the main reason for passing the horizontal movers to the BlockEdit instead of rendering them here?

jorgefilipecosta commented on packages/block-editor/src/components/block-mover/index.js:
```diff
@@ -56,12 +56,12 @@ export class BlockMover extends Component {
 					onClick={ isFirst ? null : onMoveUp }
-					icon={ upArrow }
-					label={ __( 'Move up' ) }
+					icon={ horizontalMover ? leftArrow : upArrow }
```
We would still need to customize the button labels so I guess customizing the icon does not have a significant increase in the effort.
Also, I expect the design of horizontal movers will be improved and probably it will not just be a rotation.

draganescu replied:
Yes I agree with Jorge, I think rotating won't solve much and this specifying of a separate icon adds future flexibility at little cost.

jorgefilipecosta commented on packages/block-editor/src/components/block-mover/style.scss:
```diff
@@ -1,5 +1,17 @@
+		width: auto;
+		min-height: auto;
+
+		.block-editor-block-mover__control-drag-handle {
```
We should not remove the drag handle on the horizontal movers. I guess horizontal drag&drop still does not work as expected but in that case ideally, it should be removed on the block level.

jorgefilipecosta commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -59,6 +60,7 @@ function NavigationMenuItemEdit( {
 					aria-label={ __( 'Navigation Label' ) }
 					maxRows={ 1 }
 				/>
+				{ blockMover }
```
I think blocks should not be resposible for rendeing its own movers.

jorgefilipecosta reviewed (commented):
Hi @draganescu the movers work as expected 👍 They appear in the wrong position visually for now, so it seems the most urgent fix needed.
On the dom horizontal movers are rendered inside the blocks, I think horizontal movers ideally should be rendered in the same position as vertical movers. The difference between movers should be the icons and styles, but other than that they should be as similar as possiblem.

This PR is a blocker, for the buttons block https://github.com/WordPress/gutenberg/pull/17352, for the navigation block, and I think it may also be an excellent improvement for the social links block. So, it is important to merge it soon. Let me know if I can help with any of the tasks.
I think if we mark the prop as experimental, and we fix the position (visual and dom) it is ready as a first version that we can improve after.
Any visual feedback or other requirement @mapk, @jasmussen, @kjellr, @karmatosed?

jasmussen commented:
> This PR is a blocker, for the buttons block #1735, for the navigation block, and I think it may also be an excellent improvement for the social links block. So, it is important to merge it soon. Let me know if I can help with any of the tasks.
> I think if we mark the prop as experimental, and we fix the position (visual and dom) it is ready as a first version that we can improve after.

One of the challenges with this approach is that while it adds the horizontal movers, and inline at that, it's more UI without a clear path forward toward a solution.

Zooming out a bit, those challenges are:

- There's a lot of weird margins (that I mean to refactor), making it hard to position things
- There's just a lot of UI all over the place
- In the navigation block, it's in a really small space

In https://github.com/WordPress/gutenberg/issues/17267 I outlined an idea (based on an initial concept by Matías), to let a nesting container "absorb" the UI of child blocks. This is kind of perfect for the Navigation Menu block, as it removes all those margins and provides more space for the controls.

We're obviously not going to make that in time. But a version of that was implemented in the Social Links block, where if you try it, you'll see that the nesting UI pretty well aligns with the frontend, no need for extra margins, as they are compensated for.

As an experiment, I took that further and created https://github.com/WordPress/gutenberg/pull/17627

<img src="https://user-images.githubusercontent.com/1204802/65754704-e7227700-e111-11e9-974c-d111bd255796.gif">

This concept does not have the movers in the decided upon inline place, and it assumes there's no child block toolbar other than the ellipsis menu. But it works, well, today. And the CSS to make it happen is relatively specific (and hopefully will be the foundation for solving #17267 when we get to it):

https://github.com/WordPress/gutenberg/blob/c4b057b1e892b5fb582fa89dc1c0d9f6d94b17b7/packages/block-library/src/social-links/editor.scss#L14

So what could we do as next steps to urgently fix this? I would say if the menu item child blocks don't have anything in the block toolbar other than the ellipsis menu, then we can port #17627 directly and fix the issue rather immediately. This would bring the movers to the nav menu fast. It would not replace _this_ PR, but it would allow it to get more time in the oven and provide a better mover experience with time.

draganescu committed:
implemented so as to not be a concern for the block implementer

talldan commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -53,7 +53,7 @@ function NavigationMenu( {
 				<InnerBlocks
 					allowedBlocks={ [ 'core/navigation-menu-item' ] }
-					renderAppender={ InnerBlocks.ButtonBlockAppender }
+					horizontalMover={ true }
```
In my opinion it definitely makes sense to separate the position of the mover from the orientation:
```
<InnerBlocks
  __experimentalUseInlineMovers="before|after"
  __experimentalIsHorizontalLayout
/>
```
Something like that feels right to me. We already know for the gallery block that we'll need the option of having the mover before the block edit content if that is converted to inner blocks.

draganescu replied:
yes @talldan that is exactly what I am currently implementing, a separation between the two. In thinking I arrived at the same conclusion, being that we will need all combos in various blocks simply because we style with CSS and DOM position matters.

draganescu commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -467,6 +483,7 @@ function BlockListBlock( {
 			clientId={ clientId }
 			isSelectionEnabled={ isSelectionEnabled }
 			toggleSelection={ toggleSelection }
+			blockMover={ horizontalMover ? blockMover : null }
```
this is no longer the case. following up a discussion with @jorgefilipecosta I new simply render the block mover next to the edit component, nothing is passes. It is simply respecting the same behaviour as for the current movers but enabling the styling for the intended UX.

draganescu commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -402,7 +403,8 @@ function BlockListBlock( {
 		! isPartOfMultiSelection &&
-		! isTypingWithinBlock;
+		! isTypingWithinBlock &&
+		! horizontalMover;
```
I will update this to be separate options for position and orientation.

draganescu committed:
removes useless scss variable

jasmussen commented:
Let me know if you'd like me to push some CSS from https://github.com/WordPress/gutenberg/issues/17267 into this.

draganescu commented:
@jasmussen not yet, I am refactoring this to separate concerns more.

draganescu committed:
hiding the drag handle at block level

draganescu committed:
renamed horizontalMover to moverOptions to incorporate separation of properties

draganescu committed:
rafactores the mover options

draganescu commented:
@jasmussen done refactoring, if you want to bake in this PR some CSS goodies, now is the time, until a new code review suggests edits so we can avoid conflicting work.

@jorgefilipecosta @talldan I think we're close(r) to being done here. This implementation is flexible enough to provide for all kinds of situations. Please view the updated PR description and re-review. Should we mark the options as experimental? If we do we should do it in InnerBlocks, that being the primary block developer interface, right?

talldan commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -195,6 +195,10 @@ class BlockList extends Component {
 			className,
 			blockClientIds,
 			rootClientId,
+			moverOptions = {
```
I've had feedback on one of my PRs before that it's preferable for component props not to be in an object like this. I think the reason was that memoisation of props is generally only shallow. Just thought I'd pass that along 😄

draganescu replied:
bleh, so more props?

talldan commented on packages/block-editor/src/components/block-list/block-mobile-toolbar.js:
```diff
@@ -9,11 +9,11 @@ import { ifViewportMatches } from '@wordpress/viewport';
 		<div className="editor-block-list__block-mobile-toolbar block-editor-block-list__block-mobile-toolbar">
 			<VisualEditorInserter />
-			<BlockMover clientIds={ [ clientId ] } />
+			<BlockMover clientIds={ [ clientId ] } moverOptions={ moverOptions } />
```
A similar thing happened with the PR that introduced the ButtonAppender—the props had to be passed through several layers of components.

As more customisation of inner blocks happens this might become messy.

I don't think it's something to consider for this PR, but I do wonder if an `InnerBlocks` context might be something to consider. Particular props from `InnerBlocks` could be set on the context and picked up by deeper nested components.

noisysocks replied:
Could this be a way to achieve the original API proposed by @mtias in https://github.com/WordPress/gutenberg/issues/16637#issuecomment-516410624?

That is, a block developer would specify `<InnerBlocks horizontal>` and `BlockListBlock` would detect from `InnerBlock`/`BlockList` context whether it needs to render `<BlockMover>` inside or outside the block container.

talldan replied:
Yeah, that's exactly the idea. Will save us from continually expanding those in-between components when adding these kind of customisations, at the cost of a little bit more complexity.

talldan commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -450,6 +451,21 @@ function BlockListBlock( {
 		};
 	}
 	const blockElementId = `block-${ clientId }`;
+	const { position, orientation } = moverOptions;
```
`position` and `orientation` might have to be given a more specific alias. In the context of `BlockListBlock` they could refer to lots of things.

jasmussen committed:
Initial CSS work to make the menu more manageable.

This moves to flex instead of grid, neutralizes margins, simplifies a few things.

jasmussen committed:
Make movers inline again.

jasmussen commented:
Alright, I pushed some CSS to get us closer:

![menu](https://user-images.githubusercontent.com/1204802/66039965-94342f80-e516-11e9-8eb0-867725d90989.gif)

This is _very_ complicated. The CSS is pretty hard to get right. I did this, for starters, but more work is needed:

- I copied the margin-neutralizing CSS from social links.
- Hid breadcrumbs, sibling inserters.
- Moved from CSS grid to using flex, this makes the code a fair bit more manageable.
- I made sure that the menu item has some base dimensions that perfectly match those of a menu item being edited. This means when you click a menu item to edit it, it doesn't jump around vertically. 
- I made a menu item being edited look like an input field.

There's still a ton more to be done. I think the next thing to tackle is this:

![Screenshot 2019-10-02 at 13 04 05](https://user-images.githubusercontent.com/1204802/66040209-2c321900-e517-11e9-9cb0-eab6360c02c7.png)

and this:

![Screenshot 2019-10-02 at 13 18 22](https://user-images.githubusercontent.com/1204802/66040222-30f6cd00-e517-11e9-9772-ca504e4439a9.png)

Notably, once you go into submenu territory there's a ton of extra margin I haven't yet neutralized. Doing that, I think, will help a lot. 

Even then, there's more to do. I know this PR is about the **movers**, but it's definitely touching the nav menu. And so I'm discovering these things. Probably for a separate PR, but mentioning regardless:

Do we need all those items now that we actually have movers?

![Screenshot 2019-10-02 at 12 21 08](https://user-images.githubusercontent.com/1204802/66040265-4e2b9b80-e517-11e9-8ed3-84c0f75ecd27.png)

The frontend menu block isn't styled to match the editor view at all:

![Screenshot 2019-10-02 at 12 26 19](https://user-images.githubusercontent.com/1204802/66040295-60a5d500-e517-11e9-8bb0-e6f34a67613f.png)

The frontend markup is a little duplicative:

![Screenshot 2019-10-02 at 12 27 15](https://user-images.githubusercontent.com/1204802/66040306-67344c80-e517-11e9-908c-1db5f2a0d6b9.png)

jasmussen committed:
Further improve margins for child blocks.

jasmussen commented:
Okay, pushed a bunch of CSS to polish (neutralize) the margins on nested menu items. It's getting there:

![margins](https://user-images.githubusercontent.com/1204802/66043486-918a0800-e51f-11e9-9f44-aeef6f8a7165.gif)

I do find it slightly obtuse to figure out which (+) button does what, and when. It's impressive enough that you can create infinite nesting levels in this, but I'm wondering if we shouldn't keep submenu items to something you adjust inside a popover. I worry that it will get super unwieldy once you get a lot of menu items. @shaunandrews I could use your eyes on this.

draganescu commented:
> Even then, there's more to do. I know this PR is about the movers, but it's definitely touching the nav menu.

It may as well not be touching the nav menu. I hope this PR stays "reviewable" 💃 

> The frontend menu block isn't styled to match the editor view at all:

There is a separate issue for that. #17542 

> Do we need all those items now that we actually have movers?

We don't. Actually there is an issue to remove that menu altogether #17539 

To get an overview of known issues, see #17544

pablohoneyhoney commented:
How difficult would be to hide some content while you are interacting with a child block? 
How difficult would be to move those child movers to the toolbar itself?

jorgefilipecosta commented on packages/block-editor/src/components/block-mover/index.js:
```diff
@@ -56,12 +56,12 @@ export class BlockMover extends Component {
-					label={ __( 'Move up' ) }
+					icon={ orientation === 'horizontal' ? leftArrow : upArrow }
+					label={ orientation === 'horizontal' ? __( 'Move left' ) : __( 'Move up' ) }
 					aria-describedby={ `block-editor-block-mover__up-description-${ instanceId }` }
```
Besides the aria-label the movers contain an aria description referred in aria-describedby. The aria description is generated using the function getBlockMoverDescription this function in parts of the test refers "Move down/up" I think we should make this function aware if the movers are horizontal or vertical.

jorgefilipecosta commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -511,22 +527,18 @@ function BlockListBlock( {
+					{ 'has-mover-inside': position === 'inside' },
 				) }
+			>
+				{ shouldRenderMovers && ( position === 'outside' ) && blockMover }
```
Do we really need the outside property? Can't we always render the movers inside or outside and then with CSS style them to look the way we want (inside or outside)?
If not no can we for now alway render horizontal movers inside and vertical ones outside?

The reason I would prefer to avoid inside or outside is that it seems a big contract that forces us to render the mover in a specific position. What if later we redesign the movers and all of them only make sense inside the block area? What if we move them to the toolbar?

draganescu replied:
rendering in one place is not a feasible solution as it would open a host of new issues, the most basic one is breaking the tab order, and ending with weirdly complex CSS for positioning.

I would better go for the second option and simply remove the 'position', however these two settings allow for great flexibility if we keep the current concept of movers. To answer your question, if we move the movers to the toolbar we simply deprecate the 'position' option, although we may find that we don't need any option in that case.

noisysocks replied:
I'm wondering if we need this much flexibility? When would we use `position=outside, orientation=horizontal` or `position=inside, orientation=vertical`?

talldan replied:
That's true. The positional use case that I know would be for specifying whether the mover is before the BlockEdit or after. In the gallery it'd be before, and in the nav item, after.

I think it'd be good to consider how that might be added later without actually doing it in this PR.

draganescu replied:
The problem is that if we simply ignore a position property and just render differently based on `orientation` alone we make assumptions about what horizontal means for block implementers. 

Intent is highly valuable and I do believe that by default people just want horizontal movers. However the fact that simply saying:

```
<InnerBlocks moverStyle='horizontal' /> 
```

causes the rendering of the HTML to change DOM position is something you do not expect. How could one assume that?

Offering both options is an alarm bell that tells you there is this option too to do this:

```
<InnerBlocks moverStyle='horizontal' moverPosition='beforeBlock' /> 
<InnerBlocks moverStyle='vertical' moverPosition='afterBlock' /> 
```

I will update the code to reflect this approach.

jasmussen replied:
Some of the technical aspects are a litte above me here, but it feels like we are adding a lot of options and optimizing for something that may very probably change in the future. What if we, for example, decide that the horizontal movers exist only in the block toolbar?

Another reason to be very careful in introducing options for block developers, is that we don't want block implementers to turn these controls off, as they serve important accessibility functions. 

Take the slot fill names, for example — we have intentionally kept these named after their _purpose_, not after their _location_. That way as we continue to iterate the block UI, if we have to move a slot-fill location, we don't have to rename it, and plugins will still work.

noisysocks replied:
> Take the slot fill names, for example — we have intentionally kept these named after their purpose, not after their location.

Well put! This is why I'd prefer something like `<InnerBlocks moverStyle="horizontal" />`. Where the movers appear in the DOM is an implementation detail that block authors needn't care about.

talldan replied:
> Where the movers appear in the DOM is an implementation detail that block authors needn't care about.

I don't agree with that. Being able to render it in the right place for a correct tab order is a level of control that block authors need to have. `<InnerBlocks moverStyle="horizontal" />` just won't cut it for a lot of use cases.

That said, I think the right thing to do is pick one approach, mark it as experimental, and then not remove that experimental prefix until we're tried implementing a few different blocks with these movers.

jorgefilipecosta commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -53,6 +56,7 @@ function NavigationMenu( {
 			<div className="wp-block-navigation-menu">
 				<InnerBlocks
 					allowedBlocks={ [ 'core/navigation-menu-item' ] }
+					moverOptions={ moverOptions }
```
I think for now we are still exploring an API so the public interfaces we add should be marked as __experimental.
Maybe later we decide to just have a property isHorizontal that applies to movers drag & drop etc... Marking things as experimental leaves the door open.

noisysocks replied:
Agree that we should prefix these props `__experimental` while we're figuring out the best API.

draganescu replied:
you are both right done :)

jorgefilipecosta commented on packages/block-library/src/navigation-menu/editor.scss:
```diff
@@ -1,3 +1,60 @@
+// Reduce the paddings, margins, and UI of inner-blocks.
+// @todo: eventually we may add a feature that lets a parent container absorb the block UI of a child block.
+// When that happens, leverage that instead of the following overrides.
+[data-type="core/navigation-menu"] {
```
We should coordinate the work into adding props/flags to inner blocks that provide customization options and avoid the need for these styles.

jasmussen replied:
Yep, this is ticketed in https://github.com/WordPress/gutenberg/issues/17267. But it's also necessary for us to be able to move forward, without waiting for that ticket to be worked on. What do you think?

jorgefilipecosta replied:
Yes, I think for now we can keep these styles.

jorgefilipecosta reviewed (commented):
Thank you for the updates @draganescu I left some comments but the code changes applied seem to go in the right direction 👍

jorgefilipecosta commented:
The block horizontal block movers may be applied to many different blocks so I guess to evaluate its UX it is useful to see them applied in another block.
In PR https://github.com/WordPress/gutenberg/pull/17701 it is possible to see the application of this interface to the social links block.

mapk commented:
Based on some of the discussion above, it does seem relevant to see these movers positioned similarly to Joen's social block exploration. Right now they feel a bit disconnected from the block itself. If that isn't feasible due to the overlap, it would be great to revisit the placement again once we removed the dropdown as noted in #17539. Although with the addition of a block toolbar, I'm curious to see how all these elements work together.

![Screen Shot 2019-10-02 at 2 40 21 PM](https://user-images.githubusercontent.com/617986/66084084-a1efb200-e522-11e9-9d6f-41f20a4cc2bd.png)

draganescu commented:
@mapk, @karmatosed and @shaunandrews have solved this 'disconnect' issue in their vision for the movers and in regard to how they'll look in the navigation menu block there will be a separate issue/pr, since this is meant to enable horizontal movers in a way that is reusable across blocks. What I do here with the navigation menu block is simply for showing that the API works.

draganescu commented:
@pablohoneyhoney you might want to expand on your questions, as, while they sound interesting, they are pretty open.

draganescu committed:
adds proper aliases in BlockEdit

noisysocks commented:
I tested this locally and am amazed at how far it's come. We nearly have a Navigation block! 🎉

I left some comments regarding the API in review threads above. I'd especially like us to consider if we can use context to have the API that Matias suggested in https://github.com/WordPress/gutenberg/issues/16637#issuecomment-516410624.

The one bug I see when testing is that the movers are rendered outside and vertically in _nested_ navigation menus:

<img width="687" alt="Screen Shot 2019-10-16 at 15 06 09" src="https://user-images.githubusercontent.com/612155/66887374-84444180-f026-11e9-9642-19c86335a8c1.png">

draganescu committed:
previxed options as experimental

draganescu commented:
> The one bug I see when testing is that the movers are rendered outside and vertically in nested navigation menus:

@noisysocks this is as intended because all the submenu items are vertical.

draganescu commented:
updated PR description with current issues.

noisysocks commented:
@mapk: Is this still awaiting design feedback?

draganescu committed:
RTL movers

draganescu committed:
removed the position option, marked option experimental

draganescu commented:
ok @talldan @jorgefilipecosta @noisysocks I have concluded that it is best for this PR to solve the immediate problem and have it land in master asap so it can enhance the navigation block.

Therefore:
- I have removed the new `position` option
- renamed the InnerBlock prop to `__experimentalMoverDirection`
- renamed `moverOptions` in the underlying components to `moverDirection` which can be either `(string) horizontal` or `(default) vertical`

Let me know of further things to do here :)

noisysocks commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -511,22 +526,18 @@ function BlockListBlock( {
+					{ 'has-mover-inside': moverDirection === 'horizontal' },
 				) }
+			>
+				{ shouldRenderMovers && ( moverDirection === 'vertical' ) && blockMover }
```
nit: This reads like it's a condition with no body. I'd use the ternary `?:` statement to reduce ambiguity.

```suggestion
				{ shouldRenderMovers && moverDirection === 'vertical' ? blockMover : null }
```

noisysocks commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -566,6 +577,7 @@ function BlockListBlock( {
 						{ isValid && mode === 'html' && (
 							<BlockHtml clientId={ clientId } />
 						) }
+						{ shouldRenderMovers && ( moverDirection === 'horizontal' ) && blockMover }
```
nit: Same as above.

```suggestion
						{ shouldRenderMovers && moverDirection === 'horizontal' ? blockMover : null }
```

noisysocks commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -195,6 +195,7 @@ class BlockList extends Component {
 			className,
 			blockClientIds,
 			rootClientId,
+			moverDirection = 'vertical',
```
`BlockList` is public. Should we prefix this prop with `__experimental`?

noisysocks commented on packages/block-editor/src/components/block-mover/index.js:
```diff
@@ -44,24 +44,60 @@ export class BlockMover extends Component {
 
 	render() {
-		const { onMoveUp, onMoveDown, isFirst, isLast, isDraggable, onDragStart, onDragEnd, clientIds, blockElementId, blockType, firstIndex, isLocked, instanceId, isHidden, rootClientId } = this.props;
+		const { onMoveUp, onMoveDown, orientation, isRTL, isFirst, isLast, isDraggable, onDragStart, onDragEnd, clientIds, blockElementId, blockType, firstIndex, isLocked, instanceId, isHidden, rootClientId } = this.props;
```
`BlockMover` is public. Should we prefix `orientation` with `__experimental`?

noisysocks commented on packages/block-editor/src/components/block-mover/index.js:
```diff
@@ -44,24 +44,60 @@ export class BlockMover extends Component {
 			return null;
 		}
 
+		const getArrowIcon = ( hdir ) => {
```
I think something like `moveDirection` which accepts either `'up'` or `'down'` would be a clearer argument here. That way the argument will match which direction the _block_ moves which is a constant instead of which direction the arrow faces which can change depending on `orientation` and `isRTL`.

draganescu replied:
yes, horizontal is a modification of the normal vertical movement.

noisysocks commented on packages/block-editor/src/components/block-mover/index.js:
```diff
@@ -44,24 +44,60 @@ export class BlockMover extends Component {
+					break;
+			}
+			return arrow;
+		};
```
nit: Maybe it's just me, but I find the `switch` unnecessarily confusing. Would a regular `if` be clearer?

```js
const getArrowIcon = ( moveDirection ) => {
	if ( moveDirection === 'up' ) {
		if ( orientation === 'horizontal' ) {
			return isRTL ? rightArrow : leftArrow;
		} else {
			return upArrow;
		}
	} else if ( moveDirection === 'down' ) {
		if ( orientation === 'horizontal' ) {
			return isRTL ? leftArrow : rightArrow;
		} else {
			return downArrow;
		}
	}
	return null;
};
```

draganescu replied:
mind if I copy paste this? no? I knew it!

noisysocks commented on packages/block-editor/src/components/block-mover/index.js:
```diff
@@ -44,24 +44,60 @@ export class BlockMover extends Component {
+			return arrow;
+		};
+
+		const getMovementDirection = ( mdir ) => {
```
Same again here. I think calling this `moveDirection` which accepts either `'up'` or `'down'` to match the direction that the block moves in the tree is clearer.

noisysocks commented on packages/block-editor/src/components/block-mover/index.js:
```diff
@@ -44,24 +44,60 @@ export class BlockMover extends Component {
-					label={ __( 'Move up' ) }
+					icon={ getArrowIcon( 'left' ) }
+					// translators: %s: Horizontal direction of block movement ( left, right )
+					label={ orientation === 'horizontal' ? sprintf( __( 'Move %s' ), getMovementDirection( 'left' ) ) : __( 'Move up' ) }
```
Could we move the `orientation === 'horizontal'` check into `getMovementDirection` so that there's' slightly less repetition and so that `getMovementDirection` and `getArrowIcon` behave similarly?

draganescu replied:
yes, reads much better now, thank you :)

noisysocks commented on packages/block-editor/src/components/block-mover/index.js:
```diff
@@ -79,8 +115,9 @@ export class BlockMover extends Component {
-					label={ __( 'Move down' ) }
+					icon={ getArrowIcon( 'right' ) }
+					// translators: %s: Horizontal direction of block movement ( left, right )
+					label={ orientation === 'horizontal' ? sprintf( __( 'Move %s' ), getMovementDirection( 'right' ) ) : __( 'Move down' ) }
```
Same again here about moving the `orientation === 'horizontal'` check into `getMovementDirection`.

noisysocks commented on packages/block-editor/src/components/block-mover/index.js:
```diff
@@ -44,24 +44,60 @@ export class BlockMover extends Component {
+					break;
+			}
+			return mdir;
+		};
```
nit: Similarly here, I think an `if` would be easier to grok.

noisysocks commented on packages/block-editor/src/components/block-mover/mover-description.js:
```diff
@@ -14,12 +14,30 @@ import { __, _n, sprintf } from '@wordpress/i18n';
  * @param {number}  dir           Direction of movement (> 0 is considered to be going
  *                                 down, < 0 is up).
+ * @param {string} orientation    The orientation of the block movers, vertical or horizontal.
+ * @param {boolean}   isRTL   	  True if current writing system is right to left.
```
nit: These param descriptions aren't aligned.

noisysocks commented on packages/block-editor/src/components/block-mover/mover-description.js:
```diff
@@ -14,12 +14,30 @@ import { __, _n, sprintf } from '@wordpress/i18n';
+				break;
+		}
+		return mdir;
+	};
```
Same again here about naming the parameter to `moveDirection` which accepts `'up'` or `'down'` and using an `if` instead of a `switch`.

noisysocks commented on packages/block-editor/src/components/block-mover/mover-description.js:
```diff
@@ -32,35 +50,46 @@ export function getBlockMoverDescription( selectedCount, type, firstIndex, isFir
 			type,
 			position,
-			( position + 1 )
+			orientation === 'horizontal' ? getMovementDirection( 'right' ) : 'down',
```
And again about moving `orientation` into `getMovementDirection`.

noisysocks reviewed (changes requested):
This is looking fantastic. I have only minor comments and nits now!

draganescu committed:
labeled as experimental new mober and block list props

draganescu committed:
refactored direction detection code for better readability, fixed some code alignment issues

draganescu commented:
allrighty @noisysocks check it out again please!

karmatosed commented:
My opinion here is we should remove design feedback for now. It's 'one' of a lot of navigation components and seeing them all together makes more sense than having a pause in the work there. For now, removing label as what I am seeing is what we expected.

obenland commented:
Testing with [Mayland](https://wordpress.com/theme/mayland), this is what it looked like for me:

<img width="821" alt="Screen Shot 2019-10-24 at 2 08 47 PM" src="https://user-images.githubusercontent.com/1398304/67525636-fa205c80-f667-11e9-8b3a-41aab9b87ea5.png">

Maybe we could increase styles specificity a bit?

draganescu commented:
@jasmussen I remember you added some CSS tweaks here, does @obenland 's [issue above ](https://github.com/WordPress/gutenberg/pull/16615#issuecomment-546105616) ring any bell?

jasmussen commented:
That issue doesn't immediately ring a bell. however I wonder if https://github.com/WordPress/gutenberg/issues/18104 would address this? @obenland What is the actual page structure in your screenshot? Two top level pages called Sample Page and Weird? Or is there a submenu item involved?

I'll definitely help with CSS once we figure out how where and when to fix.

draganescu commented:
Yea it's weird @jasmussen because all there's new is that .is-horizontal class that makes a horizontal mover. In @obenland 's screenshot the mover appears broken as in not being "flex". It does look too bad to be true tho :)

draganescu commented:
@obenland this seems to be a problem which is not about the css in the editor. I have tested with most of the bundled themes and as @talldan advised with the Gutenberg Starter Theme and they all work OK.

obenland commented:
Good news and bad news! 😬

I can't reproduce https://github.com/WordPress/gutenberg/pull/16615#issuecomment-546105616 anymore.

Now I'm seeing the input field with a white background (regardless of theme):

**Master:**
<img width="436" alt="Screen Shot 2019-10-28 at 10 53 27 AM" src="https://user-images.githubusercontent.com/1398304/67704085-b6806800-f971-11e9-991a-aab3d1a93512.png">

**This branch:**
<img width="441" alt="Screen Shot 2019-10-28 at 10 55 16 AM" src="https://user-images.githubusercontent.com/1398304/67704109-bf713980-f971-11e9-8887-0f8e870835f9.png">

noisysocks reviewed (approved):
LGTM! 👍 

The issue @obenland identified above exists in `master`, so let's fix that separately.

retrofox commented:
> The issue @obenland identified above exists in master, so let's fix that separately.

:+1: https://github.com/WordPress/gutenberg/pull/18172

aduth commented:
These changes appear to have made the block mover labels non-translateable.

See: #20655

draganescu commented:
Aww snap making a PR asap!

draganescu commented:
Thanks @aduth for the PR fixing this :)
